### ISSUE TITLE
Failing `\assert` for `UnionType` in `QueryPlan::analyzeQueryPlan`

### ISSUE THREAD

**Issue by jmglsn** [webonyx/graphql-php#1296]
Hi over there :)

I just ran into another assertion failure for a `UnionType`.

In `\GraphQL\Type\Definition\QueryPlan::analyzeQueryPlan` an assertion for `assert($type instanceof ObjectType || $type instanceof InterfaceType, 'proven because it must be a type with fields and was unwrapped');` fails, as it's called with a set of fieldNotes where the first field node is a `UnionType`.

This is again from a `$info->lookAhead()->queryPlan()` call similar to the one in #1293.

As you might already assume, I use a lot of union types - and try to eliminate calculating not required fields by looking into the query plan.

For what I can see right now, it seems to pass when the assertion is removed. But maybe there are other side effects that I don't see right now.

Thank you :)

**Comment by spawnia** [comment]
Thanks for your report. Hopefully I was now able to fix the issue with union types for good, I finally added a proper test for it.

While working through this, I found multiple questionable aspects of `QueryPlan` which I might change in a future major version. Feel free to leave feedback on https://github.com/webonyx/graphql-php/pull/1297 or generally.

Fix released with https://github.com/webonyx/graphql-php/releases/tag/v15.0.3.
